### PR TITLE
Refactor config code

### PR DESCRIPTION
**Pull request by tristan957**
The config code had a couple of leaky abstractions that were really quite horrible, which I can say because I wrote it :).

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

**Review by alexttx — APPROVED** [review]

File: lib/binding/kvdb_interface.c [hunk 1]
```diff
@@ -661,22 +672,24 @@ hse_kvdb_open(
...
-    err = config_deserialize_to_kvdb_rparams(conf, &params);
-    if (ev(err)) {
-        log_errx("Failed to deserialize config file for KVDB (%s) rparams", err, kvdb_home);
-        goto out;
+    if (config) {
+        err = kvdb_rparams_from_config(&params, config);
+        if (ev(err)) {
+            log_errx("Failed to deserialize config file for KVDB (%s) rparams", err, kvdb_home);
+            goto out;
+        }
```
Comment (beckerg) [inline lib/binding/kvdb_interface.c:692]: From a flow perspective I prefer function inputs to the left and outputs to the right...

File: lib/config/include/hse/config/params.h [hunk 2]
```diff
@@ -88,15 +57,15 @@ struct param_spec {
...
+    param_validator_t *ps_validate;
     /* Validates relations after conversion, validation, and updating of all data */
-    param_relation_validator_t ps_validate_relations;
+    param_relation_validator_t *ps_validate_relations;
     /* Returns a JSON string representation of the value */
-    param_stringify_t ps_stringify;
+    param_stringify_t *ps_stringify;
     /* Returns a JSON representation of the parameter */
-    param_jsonify_t ps_jsonify;
+    param_jsonify_t *ps_jsonify;
```
Comment (beckerg) [inline lib/config/include/hse/config/params.h:68]: Hard to read, better if short comment followed the decl on the same line...

**Review by beckerg — APPROVED** [review]

File: lib/binding/kvdb_interface.c (hunk 1 above)
Comment (tristan957) [reply lib/binding/kvdb_interface.c:692]: Which function are you referring to? `kvdb_rparams_from_config()`? It is an in/out parameter. It is updated from the values of the config

File: lib/config/include/hse/config/params.h (hunk 2 above)
Comment (tristan957) [reply lib/config/include/hse/config/params.h:68]: I see your point

**Pull request merged**